MemArena: Add make method to construct objects in the arena.

SolidWallOfCode opened the pull request:
Update documentation.

d2r commented on doc/developer-guide/internal-libraries/MemArena.en.rst:
```diff
@@ -14,223 +14,102 @@
+
+Other than a freeze / thaw cycle, there is no mechanism to release memory except for the destruction
+of the :class:`MemArena`. This makes :class:`MemArena` not suitable for a container that allocates
+variables sized objects (such as something like a :code:`Std::vector`). A container that allocates
```
`std::vector`

d2r commented on doc/developer-guide/internal-libraries/MemArena.en.rst:
```diff
@@ -14,223 +14,102 @@
-      Block all further allocation from any existing internal blocks. If :arg:`n` is zero then on the next allocation request a block twice as large as the current generation, otherwise the next internal block will be large enough to hold :arg:`n` bytes.
-
-   .. function:: MemArena& thaw()
+   .. function:: template <typename T> T * make(Args&& ... args)
```
```
.. function:: template <typename T, typename... Args> T* make(Args &&... args)
```

SolidWallOfCode replied:
I tried that, it causes errors in building the docs (`MemArena::make` is not found).

d2r replied:
I just tried it again, and it works:
```
   .. function:: template < typename T, typename... Args> T * make(Args&& ... args)
```

Here is my output:
```
ts:fedora:28)[~/trafficserver/doc]
linuxkit-025000000001 ∴ make html
make[1]: Entering directory '/Users/derekd/devel/trafficserver/doc'
PAPEROPT_a4="-t latex_a4" PAPEROPT_letter="-t latex_letter" PAPER="letter" PLANTUML_JAR="/Users/derekd/devel/trafficserver/.git/doc-tools/plantuml-1.2018.1/plantuml.jar" ./sbuild sphinx-build -c .  -d docbuild/doctrees -b html . docbuild/html
Running Sphinx v1.7.5
making output directory...
loading pickled environment... not yet created
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 488 source files that are out of date
updating environment: 488 added, 0 changed, 0 removed
reading sources... [ 88%] developer-guide/plugins/example-plugins/basic-authorization/implementing-the-handler-and-getting-a-handle-to-the-reading sources... [100%] preface/index.en
/Users/derekd/devel/trafficserver/doc/admin-guide/files/logging.yaml.en.rst:306: WARNING: Malformed table.
Text in column margin in table line 13.

=================== =========== ===============================================
Name                Type        Description
=================== =========== ===============================================
filename            string      The name of the logfile relative to the default
                                logging directory (set with
                                :ts:cv:`proxy.config.log.logfile_dir`).
format              string      a string with a valid named format specification.
header              string      If present, emitted as the first line of each
                                new log file.
rolling_enabled     *see below* Determines the type of log rolling to use (or
                                whether to disable rolling). Overrides
                                :ts:cv:`proxy.config.log.rolling_enabled`.
rolling_interval_sec number     Interval in seconds between log file rolling.
                                Overrides
                                :ts:cv:`proxy.config.log.rolling_interval_sec`.
rolling_offset_hr   number      Specifies an hour (from 0 to 23) at which log
                                rolling is guaranteed to align. Only has an
                                effect if RollingIntervalSec is set to greater
                                than one hour. Overrides
                                :ts:cv:`proxy.config.log.rolling_offset_hr`.
rolling_size_mb     number      Size, in megabytes, at which log files are
                                rolled.
filters             array of    The optional list of filter objects which
                    filters     restrict the individual events logged. The array
                                may only contain one accept filter.
collation_hosts     array of    If present, one or more strings specifying the
                    strings     log collation hosts to which logs should be
                                delivered, each in the form of "<ip>:<port>".
                                :ref:`admin-logging-collation` for more
                                information.
=================== =========== ===============================================
/Users/derekd/devel/trafficserver/doc/admin-guide/layer-4-routing.en.rst:: WARNING: image file not readable: admin-guide/../uml/images/l4-basic-sequence.svg
/Users/derekd/devel/trafficserver/doc/admin-guide/layer-4-routing.en.rst:86: WARNING: image file not readable: admin-guide/../uml/images/l4-example-cdn-layout.svg
/Users/derekd/devel/trafficserver/doc/admin-guide/layer-4-routing.en.rst:: WARNING: image file not readable: admin-guide/../uml/images/l4-sni-routing-seq.svg
/Users/derekd/devel/trafficserver/doc/admin-guide/logging/examples.en.rst:202: WARNING: Malformed table.
Text in column margin in table line 27.

=== ======== ==================================================================
No. Field    Description
=== ======== ==================================================================
1   cqtq     The client request timestamp in Squid format. The time of the
             client request in seconds since January 1, 1970 UTC (with
             millisecond resolution).
2   ttms     The time |TS| spent processing the client request. The number
             of milliseconds between the time the client established the
             connection with |TS| and the time |TS| sent the last byte of
             the response back to the client.
3   chi      The IP address of the client’s host machine.
4   crc/pssc The cache result code; how the cache responded to the request:
             ``HIT``, ``MISS``, and so on. Cache result codes are described
             in :ref:`admin-logging-cache-results`. The
             proxy response status code (HTTP response status code from
             |TS| to client).
5   psql     The length of the |TS| response to the client in bytes,
             including headers and content.
6   cqhm     The client request method: ``GET``, ``POST``, and so on.
7   cauc     The client request canonical URL; blanks and other characters
             that might not be parsed by log analysis tools are replaced by
             escape sequences. The escape sequence is a percentage sign
             followed by the ASCII code number of the replaced character in
             hex.
8   caun     The username of the authenticated client. A hyphen (``-``)
             means that no authentication was required.
9   phr/shn The proxy hierarchy route. The route |TS| used to retrieve the
             object.
10  psct     The proxy response content type. The object content type taken
             from the |TS| response header.
=== ======== ==================================================================
/Users/derekd/devel/trafficserver/doc/developer-guide/core-architecture/rpc.en.rst:69: WARNING: image file not readable: developer-guide/core-architecture/../../uml/images/RPC-sequence-diagram.svg
/Users/derekd/devel/trafficserver/doc/developer-guide/core-architecture/rpc.en.rst:: WARNING: image file not readable: developer-guide/core-architecture/../../uml/images/RPC-states.svg
/Users/derekd/devel/trafficserver/doc/developer-guide/plugins/example-plugins/tls_bridge.en.rst:: WARNING: image file not readable: developer-guide/plugins/example-plugins/../../../uml/images/TLS-Bridge-Plugin.svg
/Users/derekd/devel/trafficserver/doc/developer-guide/plugins/example-plugins/tls_bridge.en.rst:: WARNING: image file not readable: developer-guide/plugins/example-plugins/../../../uml/images/TLS-Bridge-Messages.svg
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
writing output... [ 88%] developer-guide/plugins/example-plugins/basic-authorization/implementing-the-handler-and-getting-a-handle-to-the-twriting output... [100%] preface/index.en
/Users/derekd/devel/trafficserver/doc/appendices/command-line/traffic_layout.en.rst:34: WARNING: unknown option: --run-root
generating indices... genindex
writing additional pages... search
copying images... [ 22%] admin-guide/monitoring/monitoring/third-party/../../../../static/images/admin/monitor/circonus/new-check-button.pncopying images... [100%] developer-guide/plugins/../../static/images/sdk/txn_sm.jpg
copying static files... done
copying extra files... done
dumping search index in English (code: en) ...
done
dumping object inventory... done
build succeeded, 10 warnings.

The HTML pages are in docbuild/html.

Build finished. The HTML pages are in docbuild/html.
make[1]: Leaving directory '/Users/derekd/devel/trafficserver/doc'
```

This gets rendered for me as:

![image](https://user-images.githubusercontent.com/905298/42226282-33f6d8f4-7ea4-11e8-8945-95dd355f17a7.png)

Maybe this requires a newer sphinx version?


I am thinking it is preferable to document the code signature and return type correctly versus documenting it incorrectly to work around a sphinx bug. But if we can confirm this is not an issue then we don't have to make that choice.

d2r replied:
@SolidWallOfCode Confirms that with sphinx 1.6.6 there is some parsing issue with this declaration. It seems we have 3 options:

1) require a minimum sphinx version
2) document the code correctly but accept the warning and lack of link
3) document the code incorrectly and have the link work.

These are in my order of preference. What do you think?

d2r commented on doc/conf.py:
```diff
@@ -177,6 +177,7 @@
                  , ('cpp:typeOrConcept', 'std::string_view')
                  , ('cpp:typeOrConcept', 'std::tuple')
                  , ('cpp:typeOrConcept', 'V') # template arguments which should be matched but aren't.
+                 , ('cpp:typeOrConcept', 'T') # template arguments which should be matched but aren't.
```
With sphinx v1.7.5, I was able to build the `html` target with `nitpicky = False` and no the `nitpicky_ignore` list.

d2r replied:
These will be removed as suggested with #3925 .

d2r commented on lib/ts/MemArena.h:
```diff
@@ -117,6 +118,21 @@ class MemArena
+      Do @b not call @c delete an object created this way - that will attempt to free the memory and break. A
+      destructor may be invoked explicitly but the point of this class is that no object in it needs to be
+      deleted, the memory will all be reclaimed when the Arena is destroyed. In general it is a bad idea
+      to make objects in the Arena that own memory that is not also in the Arena.
```
Is it possible to have `delete` calls on one of these objects fail explicitly and loudly—even if it is at run-time and not compile time?

SolidWallOfCode replied:
My experience is that it does fail quite immediately, but otherwise there's no good way to do that.

d2r replied:
OK.

d2r commented on lib/ts/unit-tests/test_MemArena.cc:
```diff
@@ -92,6 +94,20 @@ TEST_CASE("MemArena freeze and thaw", "[libts][MemArena]")
+
+    Thing(int x, std::string_view s) : ten(x), name(s) {}
+
+    Thing(std::string const &s, int x) : ten(x), name(s) {}
```
Needed here, or add a test case below?

SolidWallOfCode replied:
Possibly add a test later, not sure what I did that.

d2r replied:
Same, if it's dead code let's not add it.

d2r replied:
After the latest rework, it still seems this constructor is not used for anything. Let's remove it if we don't need it.
It looks like the others actually are used in some test.

d2r replied:
More dead code after the latest branch update. I figure it's easier just to patch it: https://github.com/SolidWallOfCode/trafficserver/pull/24

SolidWallOfCode replied:
It's not dead, these are used in the test, although I got one of them backwards. It will be updated in the next commit.

d2r commented on lib/ts/unit-tests/test_MemArena.cc:
```diff
@@ -92,6 +94,20 @@ TEST_CASE("MemArena freeze and thaw", "[libts][MemArena]")
+    Thing() {}
+    Thing(int x) : ten(x) {}
+
+    Thing(std::string const &s) : name(s) {}
```
Needed here, or add a test case below?

SolidWallOfCode replied:
I think that ends up getting used in the `PrivateThing` tests.

d2r replied:
OK let's remove and add when we actually use the code.

d2r replied:
Looks good.

d2r commented on doc/developer-guide/internal-libraries/MemArena.en.rst:
```diff
@@ -14,223 +14,110 @@
-      Number of bytes allocated out. 
-
-   .. member:: size_t next_block_size
+   .. function:: MemArena& freeze(size_t n = 0)
```
It seems like this method could be renamed to `freezeAndReserve`, since it is effectively doing more than just locking down the currently-allocated internal memory blocks.

SolidWallOfCode replied:
But there is no just `freeze` in contrast, so I think the shorter name is sufficient. There will always be _some_ next internal block size value, it is not the case calling this method create one, it only sets the value.

d2r replied:
> But there is no just freeze in contrast

But there is a `reserve`!

> There will always be some next internal block size value, it is not the case calling this method create one, it only sets the value.

Yes, this seems clear to me.

```C

size_t newSize = calculateNeededMemForNextLogic();

// Now that I know how much I need, set the internal block size for the next allocation.

arena.reserve(newSize);

// OK, now before we allocate more memory, let's free what we are not going to use.

arena.freeze(); // XXX previous reserve is clobbered here.

// Let's allocate some new memory for the stuff we want to keep around.

arena.alloc(...); // XXX Unexpected internal block size allocated here.

copyMyData(arena, ...);

// We don't need any of the memory that was previously held for the next op.

arena.thaw();
```

The above is a hypothetical example that seems reasonable but where I would not have expected the result.

With the current code, it is better to do

```C
// Prepare to clean up the old but also set the block size for the new.
arena.freeze(newSize);
```

This is what I am saying is confusing. We are passing a size parameter to freeze, which one would think would "freeze `newSize` bytes", but it's really freezing all allocated bytes and then setting some background state only to prepare for a subsequent call by different method.


I am not saying it absolutely must be changed either here or in another PR. I am only observing as a reviewer of changes to documentation that I am seeing for the first time, and this part seemed odd. If it is not that odd to others, then let's go with it.

bryancall replied:
I think it would be better to have a `freeze(bool flag = true)` method for freezing and thawing.  This is what is done with `std::strstream::freeze` https://en.cppreference.com/w/cpp/io/strstream/freeze

`reserve` and `freeze` should be separate methods.

d2r commented on doc/developer-guide/internal-libraries/MemArena.en.rst:
```diff
@@ -14,223 +14,110 @@
+:func:`MemArena::freeze` and :func:`MemArena::thaw` are provided. Calling :func:`MemArena::freeze`
+locks down all current blocks. After this call any further allocations will be in new internal
+memory blocks. The first internal block will be large enough to hold all of the currently allocated
+memory, unless a value is passed to :func:`MemArena::freeze` to set the size explicitly. This
```
* What I completely missed on the first reading is that `freeze` is in effect doing what `reserve` does in addition to locking down the internal blocks. It's really a freeze + reserve operation that happens.

  Suggest something like 
  ```
  Calling :func:`MemArena::freeze` locks down all current blocks and reserves
  memory for the next allocation. By default, :func:`MemArena::freeze` reserves
  a block that is large enough to hold all of the allocated memory that has
  been frozen.  If a value is passed to :func:`MemArena::freeze`, then it
  reserves a block of the given size.
  ```

* This may be outside the scope of this PR, but I would suggest removing the reservation functionality from this method, so that the pattern is
    * freeze
    * reserve
    * alloc
    * \<copy>
    * thaw

  rather than
    * freeze
    * alloc
    * \<copy>
    * thaw
  
  for clarity.

SolidWallOfCode commented:
Significant rewrite of the documentation.

d2r commented on doc/developer-guide/internal-libraries/MemArena.en.rst:
```diff
@@ -14,223 +14,134 @@
+memory a new internal block is allocated, each block being larger than the previous one. The size
+can be controlled using the :func:`MemArena::reserve`. In general it is not as useful to use
+:class:`MemArena` for classes that allocate their own internal memeory (such as :code:`std::string`
+or :code:`std::vector`). One container class that can be easily used is :class:`IntrusiveDList`
```
`IntrusiveDList` is not defined anywhere in the docs. Are we missing a section?

SolidWallOfCode replied:
It's waiting on #3907.

d2r commented on doc/developer-guide/internal-libraries/MemArena.en.rst:
```diff
@@ -14,223 +14,134 @@
+In normal use memory is allocated from :class:`MemArena` using :func:`MemArena::alloc` to get chunks
+of memory or classes via :func:`MemArena::make` which is effectively :code:`new`.
+:func:`MemArena::make` takes an arbitrary set of arguments which it attempts to pass to a
+constructor for the type :code:`T` after allocating memory for the object. if there isn't enough
```
`If there isn't enough` (capitalize)

d2r commented on doc/developer-guide/internal-libraries/MemArena.en.rst:
```diff
@@ -14,223 +14,134 @@
+constructor for the type :code:`T` after allocating memory for the object. if there isn't enough
+memory a new internal block is allocated, each block being larger than the previous one. The size
+can be controlled using the :func:`MemArena::reserve`. In general it is not as useful to use
+:class:`MemArena` for classes that allocate their own internal memeory (such as :code:`std::string`
```
`memory`

d2r commented on doc/developer-guide/internal-libraries/MemArena.en.rst:
```diff
@@ -14,223 +14,134 @@
+After a series of allocations which can be in different allocation blocks with (possibly) unused
+memory in those blocks (either not yet allocated or not longer in use), it can be useful to coalesce
+all of the in use memory in the arena into a single bulk allocation block. This improves memory
+efficiency and memory locality. The methods :func:`MemArena::freeze` and :func:`MemArena::thaw` are
```
> After a series of allocations which can be in different allocation blocks with (possibly) unused memory in those blocks (either not yet allocated or not longer in use), it can be useful to coalesce all of the in use memory in the arena into a single bulk allocation block. 

Suggest re-ordering:

```
After a series of allocations it can be useful to coalesce all of the in use memory in the arena into a single bulk allocation block, since memory can be in different allocation blocks with (possibly) unused memory in those blocks (either not yet allocated or not longer in use).
```

d2r commented on doc/developer-guide/internal-libraries/MemArena.en.rst:
```diff
@@ -14,223 +14,134 @@
+
+In general :class:`MemArena` will allocate memory from an internal chunk, if the allocation request
+is large (more than a memory page) and there is not enough space in the current internal block, a
+block just for that allocatoin will be created. This is useful if the purpose of :class:`MemArena`
```
`allocation`

SolidWallOfCode commented:
Used the extra constuctors on `Thing` and updated the documentation again.

d2r commented on doc/developer-guide/internal-libraries/MemArena.en.rst:
```diff
@@ -14,223 +14,140 @@
+
+Generally :class:`MemArena` is not as useful for classes that allocate their own internal memory
+(such as :code:`std::string` or :code:`std::vector`), which includes most container classes. One
+container class that can be easily used is :class:`IntrusiveDList` because the links are in the
```
`IntrusiveDList` will be defined with #3945, so this PR should be blocked on it.

SolidWallOfCode replied:
No, `IntrusiveDList` was defined with #3907. #3945 is just a minor documentation update.

d2r replied:
@SolidWallOfCode is correct. It seems there was a bug somewhere in the docs that caused the definition not to be parsed.

It seems now we have a duplicate declaration:
```
/Users/derekd/devel/trafficserver/doc/developer-guide/internal-libraries/intrusive-list.en.rst:46: WARNING: Duplicate declaration.
```

This can be fixed elsewhere.

SolidWallOfCode replied:
Yes, that's one of the things #3945 fixes.

d2r replied:
OK.

SolidWallOfCode commented:
[approve ci autest]

bryancall commented on doc/developer-guide/internal-libraries/MemArena.en.rst:
```diff
@@ -14,223 +14,137 @@
 
-      Get the current generation size. The default size of the arena is 32KB unless otherwise specified. 
+      If :arg:`flag` is :code:`true`, stop allocating from existing internal memory blocks. These
+      blocks are now "frozen". Adjust the size of the next allocation block, if necessary, to be
```
Why does freezing and un-freezing need to allocate or deallocate memory?  Shouldn't that have been done with reserve and deleting the object?

SolidWallOfCode replied:
Freezing doesn't allocate memory. Unfreezing cleans up the old memory. If it didn't do that, what would be the point of freezing at all? In theory it could be done by creating a new instance of `MemArena` and copying in to that, but that requires two allocations (one for the `MemArena`, and another for the new internal block) instead of just one, in addition to making the containing object construction more complex (i.e. it's no longer possible to declare the member simply as `MemArena`). Beyond that, one goal was to emulate the behavior of `HdrHeap` and that is how it works.

SolidWallOfCode commented:
I removed `reserve` entirely - looking at the API and use cases, I decided the use cases not involving initial allocation and freeze/thaw were not worth having the complexity of `reserve`.  As a consequence I went back to having `freeze` and `thaw` because `freeze` now needs the reservation size hint.

I changed the documentation to use the term "reserve" to mean internal allocation, terminology similar to STL. This avoids confusion because the `reserve` method is gone.

d2r commented on lib/ts/MemArena.h:
```diff
@@ -89,67 +97,74 @@ class MemArena
+   *
+   * Frozen memory is released.
+   *
+   * @return @c *this
```
Returns a reference here, not a pointer, right? Same for `freeze`?

d2r replied:
> Returns a reference here, not a pointer, right? Same for freeze?

I re-read this just now, and this is documenting the return value, not the type, and so it looks correct. Disregard the earlier questions.

SolidWallOfCode commented on lib/ts/unit-tests/test_MemArena.cc:
```diff
@@ -53,45 +58,77 @@ TEST_CASE("MemArena freeze and thaw", "[libts][MemArena]")
+  REQUIRE(arena.size() == 512);
+  REQUIRE(arena.reserved_size() >= 1024);
+
+  arena.clear();
```
It will be the size of all allocated memory. That's the size used normally  for selecting the next reserved chunk. I think I"ll add a hint to `clear` for that reason.

d2r replied:
OK let me see if I have this right:

* If we reserved _R_ bytes, then allocated _A_ < _R_ bytes (in sum) from that reservation, then call `clear()`, then call `alloc(0)`, how many bytes will be reserved?

  **Exactly _R_-_A_ bytes are reserved?**

* If we reserved _R_ bytes, then allocated _A_ > _R_ bytes (in sum; where _A_ - _R_ is less than a memory page, for simplicity) how many bytes will be reserved?

  **_R_ bytes + new a block of _S_ bytes are reserved, where _S_ >= _R_?**


Generalizing:

* When we do a `freeze` or a `clear`, the next block is as big as the **sum of what was allocated**.

* When we `alloc` more bytes (in sum) than the size of the reservation, the new block that is reserved is **at least as many bytes as was allocated** in the original block.

SolidWallOfCode replied:
* The reservation will be large enough to hold A bytes.
* R bytes in the existing reservation plus S reserved bytes such that S > A.
* Yes, the next reservation is by default large enough to hold allocated memory.
* No, the new block is large enough to hold all allocated memory, which can be in more than one block. E.g. the third reserved block is large enough to hold all the memory allocated in the first and second blocks To quote from the docs, "The default style for reservation sizing follows a Fibonacci pattern, yielding (asymptotically) a block size growth rate of phi == 1.1618... which is close to the square root of 2.".

d2r replied:
Thanks,

> The reservation will be large enough to hold A bytes.

* Then the code comment was wrong for `clear`. But I see this has been updated to refer to a hint, and it seems OK now.

> R bytes in the existing reservation plus S reserved bytes such that S > A.

* Agree

> Yes, the next reservation is by default large enough to hold allocated memory.

* OK

> No, the new block is large enough to hold all allocated memory, which can be in more than one block. 

* My statement is true I think for 2 blocks (assumed), and false for more blocks than that.

  If I am not mistaken, your statement is true for 2 blocks, true for 3 blocks, and false for more than that. The *n*th Fibonacci number is *F(n)* = *F(n-1)*+*F(n-2)*. (E.g.: Fibonacci starts like `1 1 2 3 5`. The 5th number `5` is not greater than the sum of previous numbers `1+1+2+3=7`, so its block would not be large enough to hold all allocated memory.)

  I see this is a problem because we `freeze` says that the new block will by default be large enough to hold all the memory that has been allocated so far. But this is false when there are more than 2 internal blocks that have been frozen.

SolidWallOfCode replied:
No, because the allocated memory is tracked explicitly and globally, not per block. Therefore when `frozen()` is called the hint is the total allocated memory.

d2r replied:
OK, reference to the Fibonacci behavior was removed, so this is consistent now.

SolidWallOfCode commented on lib/ts/unit-tests/test_MemArena.cc:
```diff
@@ -53,45 +58,77 @@ TEST_CASE("MemArena freeze and thaw", "[libts][MemArena]")
+  REQUIRE(arena.reserved_size() > extent); // Bigger while frozen memory is still around.
+  arena.thaw();
+  REQUIRE(arena.size() == 262144);
+  REQUIRE(arena.reserved_size() == extent); // should be identical to before freeze.
```
It should be identical.  An allocation of K bytes requires a reservation of N bytes. When the arena is frozen, then the next reservation should be enough to hold K bytes, which should also be N. Actually allocating K bytes should not require any additional reservation by definition. Therefore the extent should be identical. This does depend on the original allocation being done in one chunk.

SolidWallOfCode commented on lib/ts/unit-tests/test_MemArena.cc:
```diff
@@ -53,45 +58,77 @@ TEST_CASE("MemArena freeze and thaw", "[libts][MemArena]")
-  arena.alloc(1);
-  REQUIRE(arena.extent() >= 32000);
+  arena.alloc(0);
+  REQUIRE(arena.reserved_size() >= 32000);
```
There is a check for that down around line 234.

SolidWallOfCode replied:
Updated.

d2r replied:
Looks good.

SolidWallOfCode commented on doc/developer-guide/internal-libraries/MemArena.en.rst:
```diff
@@ -14,223 +14,143 @@
+of the copied objects will be put in the same new internal block. If this for some reason this
+sizing isn't correct a hint can be passed to :func:`MemArena::freeze` to specify a different value
+(if, for instance, there is a lot of unused memory of known size). Generally this is most useful for
+process static data that is initialized on process start. After the process start initilization, the
```
Updated.

d2r replied:
Even better.

d2r commented on lib/ts/unit-tests/test_MemArena.cc:
```diff
@@ -53,45 +58,75 @@ TEST_CASE("MemArena freeze and thaw", "[libts][MemArena]")
+
+    Thing() {}
+    Thing(int x) : ten(x) {}
+    Thing(std::string const &s) : name(s) {}
```
Would you point me to the line number that uses this constructor, `Thing(std::string const &s) : name(s) {}`?

SolidWallOfCode replied:
Looks like I missed that one, added.

d2r replied:
Looks good.

d2r commented on doc/developer-guide/internal-libraries/MemArena.en.rst:
```diff
@@ -26,211 +25,136 @@
+The arena can be **frozen** using :func:`MemArena::freeze` which locks down the currently reserved
+memory and forces the internal reservation of memory for the next allocation. By default this
+internal reservation will be the size of the frozen allocated memory. If this isn't the best value a
+a hint can be provided to the :func:`MemArena::freeze` method to specify a different value, in the
```
`If this isn't the best value a hint can be`

SolidWallOfCode replied:
Fixed

d2r commented on lib/ts/MemArena.h:
```diff
@@ -89,74 +97,84 @@ class MemArena
 
-      Empties the entire arena and deallocates all underlying memory. Next block size will be equal to the sum of all
-      allocations before the call to empty.
+      Empties the entire arena and deallocates all underlying memory. The hint for the Next reserverd block size will be
```
* `The hint for the next` (cap)
* `reserved block size will be`

SolidWallOfCode replied:
Fixed.

d2r commented on doc/developer-guide/internal-libraries/MemArena.en.rst:
```diff
@@ -26,211 +25,136 @@
+
+The arena can be **frozen** using :func:`MemArena::freeze` which locks down the currently reserved
+memory and forces the internal reservation of memory for the next allocation. By default this
+internal reservation will be the size of the frozen allocated memory. If this isn't the best value a
```
A comment in another [thread](https://github.com/apache/trafficserver/pull/3901#discussion_r205848561) has implications this statement: `By default this internal reservation will be the size of the frozen allocated memory.`

Leaving this comment here since the other thread is hidden now in my browser.

SolidWallOfCode replied:
I left a comment elsewhere, but this is true because the allocated memory size is tracked. The number of reserved blocks used is irrelevant.

d2r replied:
Looks good.

d2r commented on doc/conf.py:
```diff
@@ -167,7 +167,8 @@
 nitpicky = True
-nitpick_ignore = [
+nitpick_ignore = [ ('cpp:typeOrConcept', 'T')
+                 , ('cpp:typeOrConcept', 'Args')
```
Seems to build fine with Sphinx==1.7.5 (our minimum required version) and with Sphinx==1.7.6 (latest). We can do without these exceptions.

SolidWallOfCode replied:
I needed this for Sphinx 1.7.6 or I get warnings for these values.
```
writing output... [100%] index                                                                                                           
/home/amc/git/ats/doc/developer-guide/internal-libraries/MemArena.en.rst:127: WARNING: cpp:identifier reference target not found: T
/home/amc/git/ats/doc/developer-guide/internal-libraries/MemArena.en.rst:127: WARNING: cpp:identifier reference target not found: Args
```

SolidWallOfCode replied:
It's phase of the moon, but after several runs without errors, I got this
```
Running Sphinx v1.7.6
loading pickled environment... done
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 1 source files that are out of date
updating environment: 0 added, 1 changed, 0 removed
reading sources... [100%] developer-guide/internal-libraries/MemArena.en                                                                 
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
writing output... [100%] index                                                                                                           
/home/amc/git/ats/doc/developer-guide/internal-libraries/MemArena.en.rst:123: WARNING: cpp:identifier reference target not found: T
/home/amc/git/ats/doc/developer-guide/internal-libraries/MemArena.en.rst:123: WARNING: cpp:identifier reference target not found: Args
generating indices... genindex
writing additional pages... search
copying static files... done
copying extra files... done
dumping search index in English (code: en) ... done
dumping object inventory... done
build succeeded, 2 warnings.
```

SolidWallOfCode replied:
I need these in here in order to consistently get clean runs without warnings.

d2r replied:
Weird. I've still not seen this in my own environment. I agree if this is brittle, better to have the exceptions.

SolidWallOfCode commented:
Yes, you complained about losing changes so I decided to not squash until we were done.
